### PR TITLE
Search: Update tag filter options dynamically

### LINKED ISSUE #47068: TagFilter should load options onClick, not when the component loads
Currently the `<TagFilter />` component will load all options when the component loads -- 
```
<TagFilter isClearable tags={tags} tagOptions={getTagOptions} onChange={setTags} /> <br />
```

but we may also want the options to lazy load on click.

Can we add options so that the tag/term counts are loaded on demand?

### PR DESCRIPTION
**Pull request by nmarrs**
**What this PR does / why we need it**:
Update tag filter options dynamically based on current query results. Also tag options will not load until user interaction with tag filter (or select tag externally) vs defaulting to always loading on page load. 

Context on changes: React async select library does not have a straightforward api for updating options. Switched over to cleaner approach not using async-select to enable lazy loading. See [here](https://github.com/JedWatson/react-select/issues/1879#issuecomment-983290269) for more context.

Potentially breaking changes as this impacts all instances of TagFilter

https://user-images.githubusercontent.com/22381771/162101786-43bcce7e-0619-4295-93d9-df9097c07815.mov

**Which issue(s) this PR fixes**:

Fixes #47068

### CODE REVIEW

**Comment by nmarrs** [comment]
I tried first to go through `onMenuOpen` click (no dedicated `onClick` prop on actual async select). This resulted in user needing to click twice (once for new options to load, and second time to view updated options). Current flow means UX / flow should be relatively seamless for user vs. having to click multiple times. In a sense current flow is tied directly 1 to 1 on user action (click of a new tag to filter by / clear current tag filter etc.)

**Comment by nmarrs** [comment]
Current iteration requires user to first click on tag filter to load options and then click again to view. Not sure this is best optimization and may make more sense to just prioritize better UX over optimizing / reducing api calls (that are already partially optimized via debounce)

File: public/app/core/components/TagFilter/TagFilter.tsx [hunk 1]
```diff
@@ -45,10 +52,24 @@ export const TagFilter: FC<Props> = ({
...
+  const [selectKey, setSelectKey] = useState<string>(JSON.stringify(tags));
+  const [previousTags, setPreviousTags] = useState(tags);
 
-  const onLoadOptions = async (query: string) => {
+  useEffect(() => {
+    if (updateOptionsDynamically && previousTags !== tags) {
+      setPreviousTags(tags);
+      // Slight delay necessary to ensure query results are most up to date
+      // Key switching approach seems to be best way to ensure consistent update of tag options
+      setTimeout(() => setSelectKey(JSON.stringify(tags)), 100);
```
Comment (nmarrs) [inline public/app/core/components/TagFilter/TagFilter.tsx:68]: Seems like tag options don't dynamically update after selection unless we continue to do this key switching (in effect forcing React to re-render the select component and load up to date options)

I can continue looking into alternative approaches as this is not an ideal solution

File: public/app/core/components/TagFilter/TagFilter.tsx [hunk 2]
```diff
@@ -45,29 +49,70 @@ export const TagFilter: FC<Props> = ({
...
+    }
+  }, [onFocus, options.length, tags.length]);
+
+  useEffect(() => {
+    // Update selected tags to not include (counts) when selected externally
+    if (tags !== previousTags) {
+      setPreviousTags(tags);
+      onFocus();
+    }
+  }, [onFocus, previousTags, tags]);
```
Comment (nmarrs) [inline public/app/core/components/TagFilter/TagFilter.tsx:102]: Not super excited about these useEffects, but they are necessary to load options when user selects a tag externally (i.e. in table) / to ensure that selected tags externally don't have (counts) next to them (which get stale)

**Review by ryantxu — APPROVED** [review]
Looks great -- lets target 9 now that 8.5 is a backport process

**Pull request merged**